agent/consul/gateways/controller_gateways.go

```diff
@@ -2,201 +2,563 @@ package gateways
-		//gateway doesn't exist
-		return nil, nil
+	// construct the tuple we'll be working on to update state
+	_, bound, err := store.ConfigEntry(nil, structs.BoundAPIGateway, req.Name, req.Meta)
+	if err != nil {
+		return err
+	}
+	meta := ensureInitializedMeta(gateway, bound)
+	certificateErrors, err := meta.checkCertificates(store)
+	if err != nil {
+		return err
 	}
 
-	metaGateway.Gateway = apiGateway.(*structs.APIGatewayConfigEntry)
+	for ref, err := range certificateErrors {
+		updater.SetCondition(structs.Condition{
+			Type:               "Accepted",
+			Status:             "False",
+			Reason:             "InvalidCertificate",
+			Message:            err.Error(),
+			Resource:           &ref,
+			LastTransitionTime: &now,
+		})
+	}
+	if len(certificateErrors) > 0 {
+		updater.SetCondition(structs.Condition{
+			Type:               "Accepted",
+			Status:             "False",
+			Reason:             "InvalidCertificates",
+			Message:            "gateway references invalid certificates",
+			LastTransitionTime: &now,
+		})
+	} else {
+		updater.SetCondition(structs.Condition{
+			Type:               "Accepted",
+			Status:             "True",
+			Reason:             "Accepted",
+			Message:            "gateway is valid",
+			LastTransitionTime: &now,
+		})
+	}
 
-	boundGateway, err := r.store.GetConfigEntry(structs.BoundAPIGateway, req.Name, req.Meta)
-	if err != nil {
-		return nil, err
+	// now we bind all of the routes we can
+	updatedRoutes := []structs.ControlledConfigEntry{}
+	for _, route := range routes {
+		routeUpdater := structs.NewStatusUpdater(route)
+		_, boundRefs, bindErrors := BindRoutesToGateways([]*gatewayMeta{meta}, route)
+		// unset the old gateway binding in case it's stale
+		for _, parent := range route.GetParents() {
+			if parent.Kind == gateway.Kind && parent.Name == gateway.Name && parent.EnterpriseMeta.IsSame(&gateway.EnterpriseMeta) {
+				routeUpdater.RemoveCondition(structs.Condition{
+					Type:     "Bound",
+					Resource: &parent,
+				})
+			}
+		}
+		for _, ref := range boundRefs {
+			routeUpdater.SetCondition(structs.Condition{
+				Type:               "Bound",
+				Status:             "True",
+				Reason:             "Bound",
+				Resource:           &ref,
+				Message:            "successfully bound route",
+				LastTransitionTime: &now,
+			})
+		}
+		for reference, err := range bindErrors {
+			routeUpdater.SetCondition(structs.Condition{
+				Type:               "Bound",
+				Status:             "False",
+				Reason:             "FailedToBind",
+				Resource:           &reference,
+				Message:            err.Error(),
+				LastTransitionTime: &now,
+			})
+		}
+		if entry, updated := routeUpdater.UpdateEntry(); updated {
+			updatedRoutes = append(updatedRoutes, entry)
+		}
+	}
+
+	// first check for gateway conflicts
+	for i, listener := range meta.BoundGateway.Listeners {
+		protocol := meta.Gateway.Listeners[i].Protocol
+		switch protocol {
+		case structs.ListenerProtocolTCP:
```

Comment:
should we use a regular if statement here since the switch statement only has a single case or do we plan on adding more cases in the future?

Reply:
My thought is that we'll add more in the future, we currently don't have support for `UDPRoute` or `GRPCRoute` as defined in the upstream spec.